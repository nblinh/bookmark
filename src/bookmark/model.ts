import { dataToBookmark } from "../utils/bookmarkUtils";

export class Bookmark {
    view: string;
    url: string;
    title: string;
    author: string;
    addDate: Date;
    publicDate: Date;
    duration?: number; //in seconde
    width?: number;
    height?: number;

    constructor(view: string, url: string, title: string, author: string, publicDate: string) {
        this.view = view;
        this.url = url;
        this.title = title;
        this.author = author;
        this.addDate = new Date();
        this.publicDate = new Date(publicDate);
    }
}

export const getBookmarkInfo = async (url: string): Promise<Bookmark> => {
    if (!url.includes('vimeo.com') && !url.includes('flickr.com')) {
        throw new Error();
    }

    return fetch('https://noembed.com/embed?url=' + url)
        .then(response => response.json())
        .then(data => dataToBookmark(data))
}