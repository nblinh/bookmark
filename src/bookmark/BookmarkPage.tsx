import { FC, useState } from "react";
import BookmarkList from "./BookmarkList";
import { Bookmark, getBookmarkInfo } from "./model";

interface BookmarkPageProps {

}

const BookmarkPage: FC<BookmarkPageProps> = () => {
    const [bookmarks, setBookmarks] = useState<Bookmark[]>([]);
    const [url, setUrl] = useState("");
    const [error, setError] = useState(false);

    const handleSubmit = (evt: any) => {
        evt.preventDefault();

        getBookmarkInfo(url).then(bookmark => {
            bookmarks.push(bookmark);
            setBookmarks([...bookmarks]);
            setError(false);
        })
            .catch((error) => {
                console.error('Error:', error);
                setError(true);
            });
    }
    return (
        <div>
            <form onSubmit={handleSubmit} style={{ paddingBottom: '30px' }}>
                <label>
                    Add URL:
                    <input
                        type="text"
                        value={url}
                        data-testid="urlText"
                        onChange={e => setUrl(e.target.value)}
                    />
                </label>
                <input type="submit" value="Submit" data-testid="submitBtn" />
                <br />
                {error && <span style={{ color: 'red' }}> Please enter a valid URL!This application support only two URL type: flickr.com and vimeo.com</span>}
            </form>

            <BookmarkList bookmarks={bookmarks} setBookmarks={setBookmarks} />
        </div>
    );
}

export default BookmarkPage;