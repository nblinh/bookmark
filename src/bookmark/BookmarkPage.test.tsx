import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';
import BookmarkPage from './BookmarkPage';
import { getBookmarkInfo } from './model';


jest.mock('./model', () => ({
    getBookmarkInfo: jest.fn().mockImplementation((url: string) => Promise.resolve({}))
}));

describe('BookmarkPage', () => {
    test('should render', () => {
        const { container } = render(<BookmarkPage />);
        expect(container).toMatchSnapshot();
    });

    test('should have text field and submit button', () => {
        const { getByTestId } = render(<BookmarkPage />);
        expect(getByTestId('submitBtn')).toBeInTheDocument();
        expect(getByTestId('urlText')).toBeInTheDocument();
    });

    test('should call noembed api when click submit button', () => {
        const { getByTestId } = render(<BookmarkPage />);
        expect(getByTestId('submitBtn')).toBeInTheDocument();
        //fireEvent.click(getByTestId('submitBtn'));
        //expect(getBookmarkInfo).toHaveBeenCalledTimes(1);
    });
});