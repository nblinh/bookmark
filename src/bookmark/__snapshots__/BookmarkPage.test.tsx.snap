// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookmarkPage should render 1`] = `
<div>
  <div>
    <form
      style="padding-bottom: 30px;"
    >
      <label>
        Add URL:
        <input
          data-testid="urlText"
          type="text"
          value=""
        />
      </label>
      <input
        data-testid="submitBtn"
        type="submit"
        value="Submit"
      />
      <br />
    </form>
    <table
      data-testid="bookmarkTable"
      style="border: 1px solid black; border-collapse: collapse;"
    />
  </div>
</div>
`;
